libdrgn/iterators.c

```diff
@@ -0,0 +1,472 @@
+// Copyright (c) Meta Platforms, Inc. and affiliates.
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+#include "iterators.h"
+#include "helpers.h"
+#include "type.h"
+#include <inttypes.h>
+
+#define RADIX_TREE_ENTRY_MASK ((uint64_t)3)
+
+DEFINE_VECTOR_FUNCTIONS(frames)
+
+static struct drgn_error *
+radix_tree_iter_handle_node(struct radix_tree_iter *iter,
+			    struct drgn_object *node, uint64_t index,
+			    bool *entry_populated_ret)
+{
+	struct drgn_error *err;
+	struct radix_tree_iter_frame *frame = NULL;
+	uint64_t value;
+
+	err = drgn_object_read(node, node);
+	if (err)
+		goto out;
+	err = drgn_object_read_unsigned(node, &value);
+	if (err)
+		goto out;
+	if (((value & RADIX_TREE_ENTRY_MASK) !=
+	     iter->RADIX_TREE_INTERNAL_NODE)) {
+		// Base-case, node is NOT internal
+		err = drgn_object_bool(node, entry_populated_ret);
+		if (!err && *entry_populated_ret) {
+			iter->entry.node = *node;
```

Comment:
Eliminating the unnecessary copy definitely makes sense, but I think keeping the explicit `node` argument to the function helps readability (instead of implicitly passing it in `iter->entry.node`).

Reply:
IMO if the `node` argument is always `iter->entry.node`, then I'd be more confused about why it's a parameter.